File: problem2.cpp
// 238. Product of Array Except Self
// Time Complexity : O(m * n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english

/* 
Initialized variables to keep track of the current position in the matrix and the direction of traversal and iterating 
over each element in the matrix, adding it to the result vector while adjusting the position and direction based on the 
current location in the matrix.and returned the resulting vector containing the elements traversed in diagonal order.
*/


// Your code here along with comments explaining your approach
vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
    int m = mat.size(), n = mat[0].size(), i = 0, j = 0;
    vector<int> ans;
    bool dir = true;
    for(int k = 0; k < m*n; k++)
    {
        ans.push_back(mat[i][j]);
        if(dir)
        {
            if(j == n-1)
            {
                i++;
                dir = false;
            }
            else if(i == 0)
            {
                j++;
                dir = false;
            }
            else
            {
                i--;
                j++;
            }
        }
        else
        {
            if(i == m-1)
            {
                j++;
                dir = true;
            }
            else if(j == 0)
            {
                i++;
                dir = true;
            }
            else
            {
                i++;
                j--;
            }
            
        }
    }
    return ans;
}

File: problem-2.go
/*time complexity : O(m*n)
space complexity : O(m*n)*/

func findDiagonalOrder(mat [][]int) []int {
	n := len(mat)
	m := len(mat[0])
	i := 0
	j := 0
	dir := true //up
	var result []int
	for x := 0; x < m*n; x++ {
		result = append(result, mat[i][j])
		if dir {
			if i == 0 && j != m-1 {
				j++
				dir = false
			} else if j == m-1 {
				i++
				dir = false
			} else {
				i--
				j++
			}
		} else {
			if j == 0 && i != n-1 {
				i++
				dir = true
			} else if i == n-1 {
				j++
				dir = true
			} else {
				j--
				i++
			}
		}

	}
	return result
}
